### PR TITLE
NDRS-624: Reduce Event Size

### PR DESCRIPTION
**Pull request by darthsiroftardis**
ref: https://casperlabs.atlassian.net/browse/NDRS-614
Given the body of work still remaining and with teammates on holiday, this is issue is being split in two. Thus, currently compile time restriction of event sizes is not being enforced.

### CODE REVIEW

File: node/src/components/small_network.rs [hunk 1]
```diff
@@ -861,48 +866,46 @@ where
...
-                responder.respond(()).ignore()
-            }
-            Event::NetworkRequest {
-                req:
+                    } => {
+                        self.send_message(dest, Message::Payload(payload));
+                        responder.respond(()).ignore()
+                    }
+                    NetworkRequest::Broadcast { payload, responder } => {
+                        // We're given a message to broadcast.
```
Comment (marc-casperlabs) [inline node/src/components/small_network.rs:887]: Let's remove this comment, or keep the equivalent one for the branch above!

File: node/src/components/small_network/tests.rs [hunk 2]
```diff
@@ -143,7 +143,7 @@ impl Reactor for TestReactor {
             Event::NetworkRequest(req) => self.dispatch_event(
                 effect_builder,
                 rng,
-                Event::SmallNet(SmallNetworkEvent::from(req)),
+                Event::SmallNet(SmallNetworkEvent::NetworkRequest { req: Box::new(req) }),
```
Comment (marc-casperlabs) [inline node/src/components/small_network/tests.rs:146]: The `From` impl should be updated instead, if actually boxing here. We do not want to manually instantiate events if we can help it.

However, is this some sort of weird edge case? The `NetworkRequest` itself is wrapped using `Event::SmallNet` in what is essentially a newtype (it has only one field). Is `NetworkRequest` so big that this extra wrapping causes it to go over the maximum event size?

It would be more idiomatic to not box the request here.

File: node/src/components/storage.rs [hunk 3]
```diff
@@ -104,8 +104,13 @@ const OS_FLAGS: EnvironmentFlags = EnvironmentFlags::WRITE_MAP;
 /// Mac OS X exhibits performance regressions when `WRITE_MAP` is used.
 #[cfg(target_os = "macos")]
 const OS_FLAGS: EnvironmentFlags = EnvironmentFlags::empty();
+use static_assertions::const_assert;
+use std::mem;
+const _EVENT_SIZE: usize = mem::size_of::<Event>();
```
Comment (marc-casperlabs) [inline node/src/components/storage.rs:109]: Are we cleaning this up later? We actually want an event size restriction on the top level event, right?

If so, a component's event should probably be 95 (`repr(u8)`) or 88 bytes at most, to allow for an `enum` variant tag of the inevitable top level reactor event.

File: node/src/components/storage/tests.rs [hunk 4]
```diff
@@ -411,7 +411,7 @@ fn can_retrieve_store_and_load_deploys() {
     // Also ensure we can retrieve just the header.
     let response = harness.send_request(&mut storage, |responder| {
         StorageRequest::GetDeployHeaders {
-            deploy_hashes: smallvec![*deploy.id()],
+            deploy_hashes: Box::new(smallvec![*deploy.id()]),
```
Comment (marc-casperlabs) [inline node/src/components/storage/tests.rs:414]: This is not good, we're adding an unnecessary layer of indirection here. A `SmallVec` is conceptually already heap allocated like a `Vec` and a `Box<Vec<_>>` does make little sense in many cases.

It seems our chosen `SmallVec` is actually too big at what I assume is a size of `2`. Try reducing the size of the `SmallVec` instead, if a size of `1` would actually fit, or replace it with a plain `Vec`, which will always be 24 bytes.

File: node/src/effect/requests.rs [hunk 5]
```diff
@@ -246,14 +246,14 @@ pub enum StorageRequest {
     /// Retrieve deploys with given hashes.
     GetDeploys {
         /// Hashes of deploys to be retrieved.
-        deploy_hashes: Multiple<DeployHash>,
+        deploy_hashes: Box<Multiple<DeployHash>>,
```
Comment (marc-casperlabs) [inline node/src/effect/requests.rs:249]: As state above, `Box<Multiple<T>>` is an anti-pattern. It is basically `Box<Vec<T>>`.

File: node/src/reactor/joiner.rs [hunk 6]
```diff
@@ -213,15 +213,17 @@ impl From<StorageRequest> for Event {
 
 impl From<NetworkRequest<NodeId, Message>> for Event {
     fn from(request: NetworkRequest<NodeId, Message>) -> Self {
-        Event::SmallNetwork(small_network::Event::from(request))
+        Event::SmallNetwork(small_network::Event::NetworkRequest {
+            req: Box::new(request),
```
Comment (marc-casperlabs) [inline node/src/reactor/joiner.rs:217]: As stated above, **if** we need to do this, do not instantiate these events manually, but add a `From` `impl`, I'd say.

But please, check what's "wrong" first, causing this to no longer be a newtype-like struct.

File: node/src/reactor/joiner.rs [hunk 7]
```diff
@@ -213,15 +213,17 @@ impl From<StorageRequest> for Event {
...
     }
 }
 
 impl From<NetworkRequest<NodeId, gossiper::Message<GossipedAddress>>> for Event {
     fn from(request: NetworkRequest<NodeId, gossiper::Message<GossipedAddress>>) -> Self {
-        Event::SmallNetwork(small_network::Event::from(
-            request.map_payload(Message::from),
-        ))
+        Event::SmallNetwork(small_network::Event::NetworkRequest {
+            req: Box::new(request.map_payload(Message::from)),
```
Comment (marc-casperlabs) [inline node/src/reactor/joiner.rs:225]: Same Spiel as above, we're nesting types, but we do not want to box each time there if we can help it.

File: node/src/reactor/validator.rs [hunk 8]
```diff
@@ -540,17 +540,23 @@ impl reactor::Reactor for Reactor {
             // Requests:
             Event::NetworkRequest(req) => {
                 let event = if env::var(ENABLE_LIBP2P_ENV_VAR).is_ok() {
-                    Event::Network(network::Event::from(req))
+                    Event::Network(network::Event::NetworkRequest {
+                        request: Box::new(req),
```
Comment (marc-casperlabs) [inline node/src/reactor/validator.rs:544]: See above.

File: node/src/reactor/validator.rs [hunk 9]
```diff
@@ -540,17 +540,23 @@ impl reactor::Reactor for Reactor {
...
+                    })
                 } else {
-                    Event::SmallNetwork(small_network::Event::from(req))
+                    Event::SmallNetwork(small_network::Event::NetworkRequest { req: Box::new(req) })
```
Comment (marc-casperlabs) [inline node/src/reactor/validator.rs:547]: See above.

File: node/src/reactor/validator.rs [hunk 10]
```diff
@@ -540,17 +540,23 @@ impl reactor::Reactor for Reactor {
...
                 };
                 self.dispatch_event(effect_builder, rng, event)
             }
             Event::NetworkInfoRequest(req) => {
                 let event = if env::var(ENABLE_LIBP2P_ENV_VAR).is_ok() {
-                    Event::Network(network::Event::from(req))
+                    Event::Network(network::Event::NetworkInfoRequest {
+                        info_request: Box::new(req),
```
Comment (marc-casperlabs) [inline node/src/reactor/validator.rs:554]: See above.

File: node/src/reactor/validator.rs [hunk 11]
```diff
@@ -540,17 +540,23 @@ impl reactor::Reactor for Reactor {
...
+                    })
                 } else {
-                    Event::SmallNetwork(small_network::Event::from(req))
+                    Event::SmallNetwork(small_network::Event::NetworkInfoRequest {
+                        req: Box::new(req),
```
Comment (marc-casperlabs) [inline node/src/reactor/validator.rs:558]: See above.

**Review by marc-casperlabs — CHANGES_REQUESTED** [review]
I'd say this is pretty good! However, I believe some of the boxing warrants another look, specifically the faux-newtypes (structs with a single field) and `Box<Multiple<T>>` sites.

File: node/src/effect/requests.rs (hunk 5 above)
Comment (darthsiroftardis) [reply node/src/effect/requests.rs:258]: Using simply `Multiple` blows the compile time size limit

File: node/src/reactor/joiner.rs (hunk 7 above)
Comment (darthsiroftardis) [reply node/src/reactor/joiner.rs]: Implemented specific `impl From`

File: node/src/reactor/validator.rs (hunk 8 above)
Comment (darthsiroftardis) [reply node/src/reactor/validator.rs]: Implemented specific `impl From`

File: node/src/reactor/validator.rs (hunk 9 above)
Comment (darthsiroftardis) [reply node/src/reactor/validator.rs]: Implemented specific `impl From`

File: node/src/reactor/validator.rs (hunk 10 above)
Comment (darthsiroftardis) [reply node/src/reactor/validator.rs]: Implemented specific `impl From`

File: node/src/reactor/validator.rs (hunk 11 above)
Comment (darthsiroftardis) [reply node/src/reactor/validator.rs]: Implemented specific `impl From`

File: node/src/reactor/joiner.rs (hunk 6 above)
Comment (darthsiroftardis) [reply node/src/reactor/joiner.rs]: Implemented specific `impl From`

File: node/src/components/storage.rs (hunk 3 above)
Comment (darthsiroftardis) [reply node/src/components/storage.rs]: Added Event Specific limits for all those events which have been re-factored

File: node/src/components/small_network/tests.rs (hunk 2 above)
Comment (darthsiroftardis) [reply node/src/components/small_network/tests.rs:148]: Added a specific `From` implementation

File: node/src/effect/requests.rs (hunk 5 above)
Comment (darthsiroftardis) [reply node/src/effect/requests.rs]: Changed from `Box<Multiple<T>>`  to `Vec<T>`

File: node/src/components/storage/tests.rs (hunk 4 above)
Comment (darthsiroftardis) [reply node/src/components/storage/tests.rs]: Resolved by changing from `Mutiple` to `Vec`

File: node/src/components/storage.rs [hunk 12]
```diff
@@ -107,8 +107,13 @@ const OS_FLAGS: EnvironmentFlags = EnvironmentFlags::WRITE_MAP;
 /// Mac OS X exhibits performance regressions when `WRITE_MAP` is used.
 #[cfg(target_os = "macos")]
 const OS_FLAGS: EnvironmentFlags = EnvironmentFlags::empty();
+use static_assertions::const_assert;
+use std::mem;
+const _STORAGE_EVENT_SIZE: usize = mem::size_of::<Event>();
```
Comment (dwerner) [inline node/src/components/storage.rs:112]: Should have a newline between imports and consts.

Comment (darthsiroftardis) [reply node/src/components/storage.rs:111]: Reordered imports

File: node/src/components/network/event.rs [hunk 13]
```diff
@@ -90,17 +96,33 @@ pub enum Event<P> {
     #[from]
     NetworkRequest {
         #[serde(skip_serializing)]
-        request: NetworkRequest<NodeId, P>,
+        request: Box<NetworkRequest<NodeId, P>>,
```
Comment (marc-casperlabs) [inline node/src/components/network/event.rs:99]: Did we not manage to bring down the `NetworkRequest` down enough in size to avoid having to nest these in boxes?

File: node/src/components/network/event.rs [hunk 14]
```diff
@@ -90,17 +96,33 @@ pub enum Event<P> {
...
     },
 
-    /// A network info request made by a different component.
+    /// Incoming network info request.
     #[from]
     NetworkInfoRequest {
         #[serde(skip_serializing)]
-        info_request: NetworkInfoRequest<NodeId>,
+        info_request: Box<NetworkInfoRequest<NodeId>>,
```
Comment (marc-casperlabs) [inline node/src/components/network/event.rs:106]: Same again, can we somehow avoid having to box things twice?

**Review by marc-casperlabs — CHANGES_REQUESTED** [review]
I've looked at `NetworkRequest`, which is boxed in many events. It seems very unlikely to me that that is correct, as it would mean that `NetworkRequest` is_just_ below the threshold, with the newtype wrapper's variant discriminator putting it over.

This seems extra weird, since the message payloads and node IDs are both of varying size. Furthermore the payload is not boxed - maybe that one is worth checking out as well.

**Review by marc-casperlabs — APPROVED** [review]
Looks good to me. I'm a little surprised that our NodeIDs are so large that almost always need to be boxed but I guess we'll live with that for now :)
